nativelib/src/main/scala/scala/scalanative/unsafe/Tag.scala.gyb

```diff
@@ -244,7 +250,45 @@ object Tag {
     Tag.CArray(implicitly[Tag[T]], implicitly[Tag[N]])
   % for N in range(0, 23):
   %   targs = ", ".join(["T" + str(i) for i in range(1, N+1)] + ["R"])
-  @alwaysinline implicit def materializeCFuncPtr${N}[${targs}]: Tag[unsafe.CFuncPtr${N}[${targs}]] =
-    Tag.CFuncPtr${N}.asInstanceOf[Tag[unsafe.CFuncPtr${N}[${targs}]]]
-  % end 
+  @alwaysinline implicit def materializeCFuncPtr${N}[${targs}]: CFuncPtrTag[unsafe.CFuncPtr${N}[${targs}]] =
+		macro MacroImpl.materializeCFuncPtr${N}[${targs}]
+  % end
+
+  private class MacroImpl(val c: scala.reflect.macros.blackbox.Context) {
+    import c.universe._
+
+    val annotation = q"_root_.scala.scalanative.annotation"
+    val runtime    = q"_root_.scala.scalanative.runtime"
+    val unsafe     = q"_root_.scala.scalanative.unsafe"
+
+    val callCFunc = q"$runtime.Intrinsics.callCFuncPtr"
+    val fnPtr     = q"""
+      val _rawPtr = $runtime.toRawPtr(ptr)
+      if(loadPtr) $runtime.Intrinsics.loadRawPtr(_rawPtr)
+      else _rawPtr"""
+
+	% for N in range(0, 23):
+	%   targs 		= ", ".join(["T" + str(i) 											for i in range(1, N+1)] + ["R"])
+	%   tArgsDecl = ", ".join(["T" + str(i) + " : c.WeakTypeTag" 	for i in range(1, N+1)] + ["R: c.WeakTypeTag"])
+	%   tps	 			= ", ".join(["$T" + str(i) 											for i in range(1, N+1)] + ["$R"])
+	%   tpsDef		= "\n".join(["val T" + str(i) + " = weakTypeOf[T" + str(i) + "].dealias" for i in range(1, N+1)])
+	%   args		  = ", ".join(["$fnPtr"] + ["arg" + str(i) 									for i in range(1, N+1)])
+	%   argsDecl  = ", ".join("arg" + str(i) + ": $T" + str(i) 							for i in range(1, N+1))
+	%   CFuncPtr  = "CFuncPtr{}[{}]".format(N, tps)
+    def materializeCFuncPtr${N}[${tArgsDecl}]: c.Tree = {
```

Comment:
My main comment at this point is here: why do we need macros for this? Why can't we write user-level code to define the `Tag`s of `CFuncPtr`s?